core/src/worker/tuner/resource_based.rs

```diff
@@ -522,25 +520,74 @@ impl RealSysInfo {
             self.cur_mem_usage.store(mem, Ordering::Release);
         }
         self.cur_cpu_usage.store(cpu.to_bits(), Ordering::Release);
-        self.last_refresh.store(Instant::now());
+    }
+}
+
+/// Tracks host resource usage by refreshing metrics on a background thread.
+pub struct RealSysInfo {
+    inner: Arc<RealSysInfoInner>,
+    shutdown: Arc<AtomicBool>,
+    shutdown_handle: Mutex<Option<thread::JoinHandle<()>>>,
+}
+
+impl RealSysInfo {
+    pub(crate) fn new() -> Self {
+        let mut sys = sysinfo::System::new();
+        sys.refresh_memory();
+        let total_mem = sys.total_memory();
+        let inner = Arc::new(RealSysInfoInner {
+            sys: Mutex::new(sys),
+            cur_mem_usage: AtomicU64::new(0),
+            cur_cpu_usage: AtomicU64::new(0),
+            total_mem: AtomicU64::new(total_mem),
+        });
+        inner.refresh();
+
+        let shutdown = Arc::new(AtomicBool::new(false));
+        let thread_inner = inner.clone();
+        let thread_shutdown = shutdown.clone();
+        let handle = thread::Builder::new()
+            .name("temporal-real-sysinfo".to_string())
+            .spawn(move || {
+                const REFRESH_INTERVAL: Duration = Duration::from_millis(100);
```

Comment:
When I mentioned this to you initially, I was thinking it'd be nice to keep the behavior where we don't refresh unless required. Of course, this has the problem that you then still need to block the thing requesting the refresh while it happens.

But, that might just be a bunch of unneeded junk. Doing this every 100ms ought to be fine.